test/extensions/filters/http/csrf/csrf_filter_test.cc

```diff
@@ -0,0 +1,336 @@
+#include "common/http/header_map_impl.h"
+
+#include "extensions/filters/http/csrf/csrf_filter.h"
+
+#include "test/mocks/buffer/mocks.h"
+#include "test/mocks/http/mocks.h"
+#include "test/mocks/stats/mocks.h"
+#include "test/test_common/printers.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using testing::_;
+using testing::DoAll;
+using testing::InSequence;
+using testing::Invoke;
+using testing::NiceMock;
+using testing::Return;
+using testing::ReturnRef;
+using testing::SaveArg;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Csrf {
+
+class CsrfFilterTest : public testing::Test {
+public:
+  CsrfFilterConfigSharedPtr setupConfig() {
```

Comment:
Please make this method static to make sure it would not reference any member variables.

Reply:
Can you elaborate on what you're asking for? The proto api documentation should include the fields that are required for using this filter. From a technical standpoint there is no preferred order to these setters.

Reply:
You have implicit order of definition of member variables of the CsrfFilterTest class. For instance the filter_ must be defined after config_ which must be defined after stats_ and runtime_ members. Otherwise filter_ or config_ would be initialized using references to uninitialized objects. If someone accidentally changes that order the test could become flaky or crash.

Reply:
I'm wondering how someone could change the order. The filter requires a configuration, the configuration requires a policy, stat prefix string, stat object, and runtime object. I could be wrong but I think the test would fail if I changed the ordering of any of these. Does that sound right?

Reply:
We do this all over the place so I think this is fine.